src/trace/dma-trace.c

```diff
@@ -126,17 +129,58 @@ static enum task_state trace_work(void *data)
  */
 int dma_trace_init_early(struct sof *sof)
 {
+	int ret;
+
 	sof->dmat = rzalloc(SOF_MEM_ZONE_SYS_SHARED, 0, SOF_MEM_CAPS_RAM, sizeof(*sof->dmat));
+	if (!sof->dmat) {
+		ret = -ENOMEM;
+		goto err;
+	}
+
+	ret = dma_trace_buffer_init(sof->dmat);
+	if (ret)
+		goto err;
+
 	dma_sg_init(&sof->dmat->config.elem_array);
 	spinlock_init(&sof->dmat->lock);
 
 	ipc_build_trace_posn(&sof->dmat->posn);
 	sof->dmat->msg = ipc_msg_init(sof->dmat->posn.rhdr.hdr.cmd,
 				      sizeof(sof->dmat->posn));
-	if (!sof->dmat->msg)
-		return -ENOMEM;
+	if (!sof->dmat->msg) {
+		ret = -ENOMEM;
+		goto err;
+	}
+
+	/* It should be the very first sent log for easily identification. */
+	mtrace_printf(LOG_LEVEL_INFO,
+		      "FW ABI 0x%x DBG ABI 0x%x tag " SOF_GIT_TAG " src hash 0x%08x (ldc hash "
+		      META_QUOTE(SOF_SRC_HASH) ")",
+		      SOF_ABI_VERSION, SOF_ABI_DBG_VERSION, SOF_SRC_HASH);
+
+	tr_info(&dt_tr,
+		"FW ABI 0x%x DBG ABI 0x%x tag " SOF_GIT_TAG " src hash 0x%08x (ldc hash "
+		META_QUOTE(SOF_SRC_HASH) ")",
+		SOF_ABI_VERSION, SOF_ABI_DBG_VERSION, SOF_SRC_HASH);
 
 	return 0;
+
+err:
+	mtrace_printf(LOG_LEVEL_ERROR,
+		      "dma_trace_init_early() failed: %d", ret);
+
+	if (sof->dmat && sof->dmat->dmatb.addr)
+		dma_trace_buffer_free(sof->dmat);
+
+	if (false)
+		/* This crashes for some unknown reason. Circular
+		 * initialization dependencies?
+		 */
+		rfree(sof->dmat);
```

Comment:
This looks a bit weird and something to address before merging.

Reply:
I don't like this either but I'm very reluctant to spend a few extra days debugging the probably half-initialized memory allocator that I don't know anything about yet. With this, #4327 and others I've already spent an inordinate amount of time fixing issues in a tracing solution that we want to deprecate eventually. @lgirdwood, @keyonjie any clue about this crash or any name of an expert in this area?

The impact of this issue seems small: a tiny amount of memory wasted in the very unusual case where DMA tracing fails to initialize here = itself a much "bigger problem"! In fact I bet no one would have noticed if I had simply omitted this `rfree()` entirely or even worse: not tested error handling (typical). Now that I narrowed down this crash, I think it's better to have this "strange"/commented out code here to save a lot of time to whoever would try to add it back if I it were missing. It's not like we can use gdb, can we?

Thoughts?

Reply:
Lets not waste time on fixing this, lest spend our time getting the Zephyr DMA backend working.

Reply:
I fundamentally disagree on the decision-making process @lgirdwood. Both memory allocation and trace issues are now being pushed under the rug with a let's use Zephyr argument, that's a shortcut that needs a wider discussion.

The direction has to be one of 

a) we have no plans to have Zephyr used in public and Chrome releases on APL..TGL, so we need to fix the trace and memory allocator
b) we formally decide at the TSG that we ditch the existing SOF with XTOS baseline to use Zephyr only.

I am not against Zephyr, far from it, but we need a formal plan for a transition.

Reply:
Direction is b. Makes no sense maintaining two RTOSes. The trace/memory issue are not critical yet and we have also been living with them for a while.
I will schedule TSC for immediately after v1.8 to discuss transition plan.